File: src/main/java/net/arvandor/numinoustreasury/command/numinouslog/NuminousLogCommand.java
package net.arvandor.numinoustreasury.command.numinouslog;

import static net.md_5.bungee.api.ChatColor.RED;

import net.arvandor.numinoustreasury.NuminousTreasury;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class NuminousLogCommand implements CommandExecutor, TabCompleter {

    private final NuminousLogViewCommand numinousLogViewCommand;
    private final NuminousLogAddCommand numinousLogAddCommand;

    private final List<String> viewAliases = List.of("view", "show");
    private final List<String> addAliases = List.of("add", "append");

    private final List<String> subcommands = new ArrayList<>() {{
        addAll(viewAliases);
        addAll(addAliases);
    }};

    public NuminousLogCommand(NuminousTreasury plugin) {
        numinousLogViewCommand = new NuminousLogViewCommand(plugin);
        numinousLogAddCommand = new NuminousLogAddCommand(plugin);
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (args.length < 1) {
            sender.sendMessage(RED + "Usage: /numinouslog [view|add]");
            return true;
        }
        if (viewAliases.contains(args[0].toLowerCase())) {
            return numinousLogViewCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (addAliases.contains(args[0].toLowerCase())) {
            return numinousLogAddCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else {
            sender.sendMessage(RED + "Usage: /numinouslog [view|add]");
            return true;
        }
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        if (args.length < 1) {
            return subcommands;
        } else if (args.length == 1) {
            return subcommands.stream().filter(subcommand -> subcommand.startsWith(args[0].toLowerCase())).toList();
        } else {
            if (viewAliases.contains(args[0].toLowerCase())) {
                return numinousLogViewCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (addAliases.contains(args[0].toLowerCase())) {
                return numinousLogAddCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else {
                return List.of();
            }
        }
    }
}

File: src/main/kotlin/net/arvandor/numinoustreasury/command/numinouslog/NuminousLogCommand.kt
package net.arvandor.numinoustreasury.command.numinouslog

import net.arvandor.numinoustreasury.NuminousTreasury
import net.md_5.bungee.api.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter

class NuminousLogCommand(plugin: NuminousTreasury) : CommandExecutor, TabCompleter {
    private val numinousLogViewCommand = NuminousLogViewCommand(plugin)
    private val numinousLogAddCommand = NuminousLogAddCommand(plugin)

    private val viewAliases = listOf("view", "show")
    private val addAliases = listOf("add", "append")

    private val subcommands = viewAliases + addAliases

    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): Boolean {
        if (args.isEmpty()) {
            sender.sendMessage(ChatColor.RED.toString() + "Usage: /numinouslog [view|add]")
            return true
        }
        if (viewAliases.contains(args[0].lowercase())) {
            return numinousLogViewCommand.onCommand(
                sender,
                command,
                label,
                args.drop(1).toTypedArray(),
            )
        } else if (addAliases.contains(args[0].lowercase())) {
            return numinousLogAddCommand.onCommand(
                sender,
                command,
                label,
                args.drop(1).toTypedArray(),
            )
        } else {
            sender.sendMessage(ChatColor.RED.toString() + "Usage: /numinouslog [view|add]")
            return true
        }
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): List<String>? {
        return when {
            args.size < 1 -> subcommands
            args.size == 1 -> subcommands.filter { subcommand: String -> subcommand.startsWith(args[0].lowercase()) }
            else ->
                when (args[0].lowercase()) {
                    in viewAliases ->
                        numinousLogViewCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    in addAliases ->
                        numinousLogAddCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    else -> listOf()
                }
        }
    }
}
